File: product-brain/src/product_brain/index/write.py
from __future__ import annotations

from dataclasses import asdict
from datetime import datetime
from pathlib import Path

import yaml

from ..models import TicketRecord


_MANUAL_SENTINEL = "<!-- manual: do not overwrite below this line -->"


def _to_iso(dt) -> str:
    if dt is None:
        return ""
    if isinstance(dt, datetime):
        return dt.isoformat()
    return str(dt)


def _front_matter_dict(record: TicketRecord) -> dict:
    return {
        "ticket": record.ticket,
        "title": record.title,
        "type": record.type,
        "status": record.status,
        "first_commit": _to_iso(record.first_commit),
        "last_commit": _to_iso(record.last_commit),
        "shas": record.shas,
        "prs": record.prs,
        "authors": record.authors,
        "files": [
            {"path": f.path, "change": f.change, "loc_added": f.loc_added, "loc_removed": f.loc_removed}
            for f in record.files
        ],
        "symbols": record.symbols,
        "related_tickets": record.related_tickets,
        "reverted_by": record.reverted_by,
        "linked_bugs": record.linked_bugs,
        "loc_added": record.loc_added,
        "loc_removed": record.loc_removed,
        "duration_days": round(record.duration_days, 2),
        "pr_open_to_merge_days": (round(record.pr_open_to_merge_days, 2) if record.pr_open_to_merge_days is not None else None),
        "manual_sections": record.manual_sections,
        "test_cases": [
            {
                "id": c.id,
                "title": c.title,
                "automation": c.automation,
                "type": c.type,
                "suite": c.suite,
                "linked_tickets": c.linked_tickets,
                "last_status": c.last_status,
                "last_run": _to_iso(c.last_run),
                "recent_failures": c.recent_failures,
                "url": c.url,
            }
            for c in record.test_cases
        ],
        "coverage_gaps": [
            {"edge": g.edge, "edge_source": g.edge_source, "rationale": g.rationale}
            for g in record.coverage_gaps
        ],
    }


def _render_bullets(bullets) -> str:
    out = []
    for b in bullets:
        out.append(f"- {b.text}\n  source: {b.source}")
    return "\n".join(out) if out else "_(none)_"


def render(record: TicketRecord) -> str:
    front = _front_matter_dict(record)
    front_yaml = yaml.safe_dump(front, sort_keys=False).strip()

    decisions = "\n".join(f"- {d}" for d in record.key_decisions) or "_(none)_"
    edges = _render_bullets(record.edge_cases_handled)
    gaps = _render_bullets(record.known_gaps)
    qa_edges = _render_bullets(record.qa_edges)
    stability = "\n".join(f"- {s}" for s in record.stability_signals) or "_(none)_"
    coverage = "\n".join(
        f"- {g.edge}\n  source: {g.edge_source}\n  rationale: {g.rationale}"
        for g in record.coverage_gaps
    ) or "_(none)_"
    manual = record.manual_body or f"\n{_MANUAL_SENTINEL}\n## Edge cases (manual)\n\n<!-- Engineers may add hand-written edge cases here. Not LLM-managed. -->\n"

    return f"""---
{front_yaml}
---

## What shipped

{record.what_shipped or "_(no signals)_"}

## Key decisions

{decisions}

## Edge cases handled

{edges}

## Known gaps

{gaps}

## QA-verified edges

{qa_edges}

## Stability signals

{stability}

## Coverage gaps

{coverage}

{manual.strip()}
"""


def write_record(repo_path: Path, record: TicketRecord) -> Path:
    base = repo_path / ".product-brain" / "tickets"
    base.mkdir(parents=True, exist_ok=True)
    p = base / f"{record.ticket}.md"
    p.write_text(render(record))
    return p

File: product-brain-ts/src/records/write.ts
// Port target: ../product-brain/src/product_brain/index/write.py
import { mkdirSync, writeFileSync } from "node:fs";
import path from "node:path";
import yaml from "js-yaml";
import type { EdgeCaseBullet, Manifest, TicketRecord } from "../models.js";

const MANUAL_SENTINEL = "<!-- manual: do not overwrite below this line -->";

function toIso(dt: Date | string | undefined): string {
  if (!dt) return "";
  if (dt instanceof Date) return dt.toISOString();
  return String(dt);
}

function frontMatterDict(r: TicketRecord): Record<string, unknown> {
  return {
    ticket: r.ticket,
    title: r.title,
    type: r.type,
    status: r.status,
    first_commit: toIso(r.firstCommit),
    last_commit: toIso(r.lastCommit),
    shas: r.shas,
    prs: r.prs,
    authors: r.authors,
    files: r.files.map((f) => ({
      path: f.path,
      change: f.change,
      loc_added: f.loc_added ?? 0,
      loc_removed: f.loc_removed ?? 0,
    })),
    symbols: r.symbols,
    related_tickets: r.relatedTickets,
    reverted_by: r.revertedBy,
    linked_bugs: r.linkedBugs,
    loc_added: r.locAdded,
    loc_removed: r.locRemoved,
    duration_days: Math.round(r.durationDays * 100) / 100,
    pr_open_to_merge_days:
      r.prOpenToMergeDays != null ? Math.round(r.prOpenToMergeDays * 100) / 100 : null,
    manual_sections: r.manualSections,
    test_cases: r.testCases.map((c) => ({
      id: c.id,
      title: c.title,
      automation: c.automation,
      type: c.type,
      suite: c.suite,
      linked_tickets: c.linkedTickets,
      last_status: c.lastStatus ?? null,
      last_run: toIso(c.lastRun),
      recent_failures: c.recentFailures,
      url: c.url,
    })),
    coverage_gaps: r.coverageGaps.map((g) => ({
      edge: g.edge,
      edge_source: g.edgeSource,
      rationale: g.rationale,
    })),
  };
}

function renderBullets(bullets: EdgeCaseBullet[]): string {
  if (bullets.length === 0) return "_(none)_";
  return bullets.map((b) => `- ${b.text}\n  source: ${b.source}`).join("\n");
}

export function render(r: TicketRecord): string {
  const frontYaml = yaml.dump(frontMatterDict(r), { sortKeys: false, lineWidth: -1 }).trim();
  const decisions = r.keyDecisions.length
    ? r.keyDecisions.map((d) => `- ${d}`).join("\n")
    : "_(none)_";
  const edges = renderBullets(r.edgeCasesHandled);
  const gaps = renderBullets(r.knownGaps);
  const qaEdges = renderBullets(r.qaEdges);
  const stability = r.stabilitySignals.length
    ? r.stabilitySignals.map((s) => `- ${s}`).join("\n")
    : "_(none)_";
  const coverage = r.coverageGaps.length
    ? r.coverageGaps
        .map((g) => `- ${g.edge}\n  source: ${g.edgeSource}\n  rationale: ${g.rationale}`)
        .join("\n")
    : "_(none)_";
  const manual =
    r.manualBody ||
    `\n${MANUAL_SENTINEL}\n## Edge cases (manual)\n\n<!-- Engineers may add hand-written edge cases here. Not LLM-managed. -->\n`;

  return `---
${frontYaml}
---

## What shipped

${r.whatShipped || "_(no signals)_"}

## Key decisions

${decisions}

## Edge cases handled

${edges}

## Known gaps

${gaps}

## QA-verified edges

${qaEdges}

## Stability signals

${stability}

## Coverage gaps

${coverage}

${manual.trim()}
`;
}

export function writeRecord(brainRoot: string, record: TicketRecord): string {
  if (!record.repo) {
    throw new Error("TicketRecord.repo must be set before writeRecord");
  }
  const base = path.join(brainRoot, "repos", record.repo, "tickets");
  mkdirSync(base, { recursive: true });
  const p = path.join(base, `${record.ticket}.md`);
  writeFileSync(p, render(record));
  return p;
}

export function writeManifest(brainRoot: string, manifest: Manifest): void {
  const p = path.join(brainRoot, "repos", manifest.repo, "manifest.md");
  mkdirSync(path.dirname(p), { recursive: true });
  const front = {
    repo: manifest.repo,
    ticket_regex: manifest.ticketRegex,
    workflow: manifest.workflow,
    languages: manifest.languages,
    entry_points: manifest.entryPoints,
    owners_file: manifest.ownersFile,
    ignore_paths: manifest.ignorePaths,
    mega_file_threshold: manifest.megaFileThreshold,
    last_indexed_sha: manifest.lastIndexedSha,
    index_cutoff_date: manifest.indexCutoffDate,
  };
  const frontYaml = yaml.dump(front, { sortKeys: false, lineWidth: -1 }).trim();
  writeFileSync(p, `---\n${frontYaml}\n---\n\n${manifest.body}\n`);
}
